core/config/config.go

```diff
@@ -144,41 +145,61 @@ func GetEnvironmentPrefix() string {
 // ProviderFunc is used to create config providers on configuration initialization
 type ProviderFunc func() (ConfigurationProvider, error)
 
+// DynamicProviderFunc is used to create config providers on configuration initialization
+type DynamicProviderFunc func(config ConfigurationProvider) (ConfigurationProvider, error)
+
 // RegisterProviders registers configuration providers for the global config
 func RegisterProviders(providerFuncs ...ProviderFunc) {
 	_cpMux.Lock()
 	defer _cpMux.Unlock()
-	_configProviderFuncs = append(_configProviderFuncs, providerFuncs...)
+	_staticProviderFuncs = append(_staticProviderFuncs, providerFuncs...)
+}
+
+// RegisterDynamicProviders registers dynamic config providers for the global config
+func RegisterDynamicProviders(dynamicProviderFuncs ...DynamicProviderFunc) {
+	_cpMux.Lock()
+	defer _cpMux.Unlock()
+	_dynamicProviderFuncs = append(_dynamicProviderFuncs, dynamicProviderFuncs...)
 }
 
 // Providers should only be used during tests
 func Providers() []ProviderFunc {
-	return _configProviderFuncs
+	return _staticProviderFuncs
 }
 
 // UnregisterProviders clears all the default providers
 func UnregisterProviders() {
 	_cpMux.Lock()
 	defer _cpMux.Unlock()
-	_configProviderFuncs = nil
+	_staticProviderFuncs = nil
 }
 
 // InitializeGlobalConfig initializes the ConfigurationProvider for use in a service
 func InitializeGlobalConfig() {
 	_setupMux.Lock()
 	defer _setupMux.Unlock()
```

Comment:
shouldn't this be the same mutex as config provider registration? i think you might have race conditions otherwise